Pcap++/src/PcapLiveDevice.cpp

```diff
@@ -1056,33 +1063,118 @@ void PcapLiveDevice::setDefaultGateway()
-	if (ifaceInfo == "")
-	{
-		PCPP_LOG_DEBUG("Error retrieving default gateway address: couldn't get netstat output");
+#elif defined(__APPLE__)
+
+	//route message struct for communication in APPLE device
+	struct BSDRoutingMessage{
+		struct	rt_msghdr header;
+		char	messageSpace[512];
+	};
+
+	struct BSDRoutingMessage routingMessage;
+	char* spacePtr = routingMessage.messageSpace;
+	// It creates a raw socket that can be used for routing-related operations
+	int sockfd = socket(PF_ROUTE, SOCK_RAW, 0);
+	if (sockfd < 0) {
+		PCPP_LOG_ERROR("Error retrieving default gateway address: couldn't get open routing socket");
+		return ;
+	}
+	memset(reinterpret_cast<char*>(&routingMessage), 0, sizeof(routingMessage));
+	routingMessage.header.rtm_msglen = sizeof(struct rt_msghdr);
+	routingMessage.header.rtm_version = RTM_VERSION;
+	routingMessage.header.rtm_type = RTM_GET;
+	routingMessage.header.rtm_addrs = RTA_DST | RTA_NETMASK | RTA_IFP;
+	routingMessage.header.rtm_flags = RTF_UP | RTF_GATEWAY | RTF_STATIC;
+
+	struct	sockaddr_in so_dst = {0} , so_mask = {0};
+	struct	sockaddr_dl so_ifp = {0};
+
+	size_t len = sizeof(sockaddr_in);
+	bcopy(reinterpret_cast<char*>(&so_dst), spacePtr, len);
+	spacePtr += len;
+	bcopy(reinterpret_cast<char*>(&so_mask), spacePtr, len);
+	spacePtr += len;
+	routingMessage.header.rtm_msglen += 2*len ;
+	len = sizeof(sockaddr_dl);
+	bcopy(reinterpret_cast<char*>(&so_ifp), spacePtr, len);
+	routingMessage.header.rtm_msglen += len ;
+
+
+	if (write(sockfd, reinterpret_cast<char*>(&routingMessage), routingMessage.header.rtm_msglen) < 0) {
+		PCPP_LOG_ERROR("Error retrieving default gateway address: couldn't write into the routing socket");
 		return;
 	}
 
-	// remove the word "default"
-	ifaceInfo.erase(0, 7);
-
-	// remove spaces
-	while (ifaceInfo.at(0) == ' ')
-		ifaceInfo.erase(0,1);
-
-	// erase string after gateway IP address
-	ifaceInfo.resize(ifaceInfo.find(' ', 0));
+	// Read the response from the route socket
+	if (read(sockfd, reinterpret_cast<char*>(&routingMessage), sizeof(routingMessage)) < 0) {
+		PCPP_LOG_ERROR("Error retrieving default gateway address: couldn't read from the routing socket");
+		return;
+	}
 
+	struct in_addr  *gateAddr = nullptr;
+	struct sockaddr_dl *ifp = nullptr;
+	struct sockaddr *sa = nullptr;
+	spacePtr = (reinterpret_cast<char*>(&routingMessage.header+1));
+	for (int i = 1; i != 0; i <<= 1)
+	{
+		if (i & routingMessage.header.rtm_addrs)
+		{
+			sa =  reinterpret_cast<sockaddr *>(spacePtr);
+			switch (i)
+			{
+			case RTA_GATEWAY:
+				gateAddr = internal::sockaddr2in_addr(sa);
+				break;
+			case RTA_IFP:
+				if (sa->sa_family == AF_LINK &&
+						reinterpret_cast<sockaddr_dl*>(sa)->sdl_nlen)
+					ifp = reinterpret_cast<sockaddr_dl *>(sa);
+				break;
+			}
+			// Make sure the increment is the nearest multiple of the size of uint32_t
+			spacePtr += sa->sa_len > 0 ? (1 + (((sa->sa_len) - 1) | (sizeof(uint32_t) - 1))) : sizeof(uint32_t);
+		}
```

Comment:
I try to keep this function inside the the `#elif defined(__APPLE__)` Marco, since we only used this function once here. So I used a lambda function instead.